### PR TITLE
Build Env Groups at Build Time

### PR DESCRIPTION
**Pull request by sdess09**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Other (please describe):

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] If it's a backend change, tests for the changes have been added and `go test ./...` runs successfully from the root folder.
- [ ] If it's a frontend change, Prettier has been run
- [ ] Docs have been reviewed and added / updated if needed

### CODE REVIEW

File: cli/cmd/stack/apply.go [hunk 1]
```diff
@@ -172,9 +172,13 @@ func createStackConf(client *api.Client, app *Application, stackName string, pro
 	}
```
Comment (stefanmcshane) [inline cli/cmd/stack/apply.go:172]: @ferozemohideen should this path be renamed away from stacks also?

File: cli/cmd/stack/apply.go [hunk 2]
```diff
@@ -255,6 +259,58 @@ func convertToBuild(porterApp *types.PorterApp) Build {
...
+	namespace := fmt.Sprintf("porter-stack-%s", stackName)
+	release, err := client.GetRelease(
+		context.Background(),
+		projectID,
+		clusterID,
+		namespace,
+		stackName,
+	)
+
+	if err == nil && release != nil {
```
Comment (stefanmcshane) [inline cli/cmd/stack/apply.go:275]: Check that the error is not nil first, always.

File: cli/cmd/stack/apply.go [hunk 3]
```diff
@@ -255,6 +259,58 @@ func convertToBuild(porterApp *types.PorterApp) Build {
...
+						}
+					}
+				}
+			}
+		}
+	}
+
+	if envGroups == nil {
+		return envVarsGroupStringMap
+	} else {
```
Comment (stefanmcshane) [inline cli/cmd/stack/apply.go:292]: No need for else here as its implicit.

File: cli/cmd/stack/apply.go [hunk 4]
```diff
@@ -255,6 +259,58 @@ func convertToBuild(porterApp *types.PorterApp) Build {
...
+		envGroupList, err := client.ListEnvGroups(context.Background(),
```
Comment (stefanmcshane) [inline cli/cmd/stack/apply.go:293]: shouldnt have context.Background in any PR unless absolutely necessary

File: cli/cmd/stack/apply.go [hunk 5]
```diff
@@ -255,6 +259,58 @@ func convertToBuild(porterApp *types.PorterApp) Build {
...
+			projectID,
+			clusterID)
+		if err == nil {
```
Comment (stefanmcshane) [inline cli/cmd/stack/apply.go:296]: Handle error condition first, always.

File: api/types/porter_app.go [hunk 6]
```diff
@@ -143,3 +143,15 @@ type ServiceDeploymentMetadata struct {
 	// Type is the type of the service - one of web, worker, or job
 	Type string `json:"type"`
 }
+type ListEnvironmentGroupsResponse struct {
+	EnvironmentGroups []EnvironmentGroupListItem `json:"environment_groups,omitempty"`
+}
+
+type EnvironmentGroupListItem struct {
```
Comment (stefanmcshane) [inline api/types/porter_app.go:150]: Dont forget comments on exported structs and struct fields

File: api/client/porter_app.go [hunk 7]
```diff
@@ -127,3 +127,22 @@ func (c *Client) CreateOrUpdatePorterAppEvent(
 
 	return *resp, err
 }
+
+// List Env Groups (List all Env Groups for a given cluster)
```
Comment (stefanmcshane) [inline api/client/porter_app.go:131]: Exported comment names should be of the same name as the function .i.e //ListEnvGroups

**Review by ferozemohideen — APPROVED** [review]

**Pull request merged**